diff -ur a/xpra/scripts/main.py b/xpra/scripts/main.py
--- a/xpra/scripts/main.py	2009-11-10 22:42:21.000000000 +0700
+++ b/xpra/scripts/main.py	2009-11-10 21:56:15.000000000 +0700
@@ -199,10 +199,13 @@
 
 def run_client(parser, opts, extra_args):
     from xpra.client import XpraClient
-    sock = connect(pick_display(parser, opts, extra_args))
+    try:
+        sock = connect(pick_display(parser, opts, extra_args))
+    except socket.error, e:
+        sys.stdout.write("Failed to connect to socket: %s\n" % e)
+        return
     if opts.compression_level < 0 or opts.compression_level > 9:
         parser.error("Compression level must be between 0 and 9 inclusive.")
-    app = XpraClient(sock, opts.compression_level)
     sys.stdout.write("Attached (press Control-C to detach)\n")
     app.run()

