diff -ur a/wimpiggy/error.py b/wimpiggy/error.py
--- a/wimpiggy/error.py	2009-11-08 19:17:54.000000000 +0700
+++ b/wimpiggy/error.py	2009-11-10 18:27:30.000000000 +0700
@@ -26,11 +26,9 @@
 # super-fast connections to the X server, everything running on fast
 # computers... does being this careful to avoid sync's actually matter?)
 
-__all__ = ["XError", "trap"]
-
 import sys
 
-import gtk.gdk
+import gtk.gdk as _gdk
 
 from wimpiggy.log import Logger
 log = Logger()
@@ -59,16 +57,16 @@
 
     def _enter(self):
         assert self.depth >= 0
-        gtk.gdk.error_trap_push()
+        _gdk.error_trap_push()
         self.depth += 1
 
     def _exit(self, need_sync):
         assert self.depth >= 0
         self.depth -= 1
         if self.depth == 0 and need_sync:
-            gtk.gdk.flush()
+            _gdk.flush()
         # This is a Xlib error constant (Success == 0)
-        error = gtk.gdk.error_trap_pop()
+        error = _gdk.error_trap_pop()
         if error:
             if error in _exc_for_error:
                 raise _exc_for_error[error](error)
diff -ur a/wimpiggy/prop.py b/wimpiggy/prop.py
--- a/wimpiggy/prop.py	2009-11-08 19:17:54.000000000 +0700
+++ b/wimpiggy/prop.py	2009-11-10 20:11:51.000000000 +0700
@@ -15,8 +15,8 @@
 from wimpiggy.lowlevel import \
      XGetWindowProperty, XChangeProperty, PropertyError, \
      get_xatom, get_pyatom, get_xwindow, get_pywindow, const, \
-     get_display_for, premultiply_argb_in_place
-from wimpiggy.error import *
+     premultiply_argb_in_place
+from wimpiggy.error import trap, XError
 from wimpiggy.log import Logger
 log = Logger()
 
diff -ur a/wimpiggy/test_error.py b/wimpiggy/test_error.py
--- a/wimpiggy/test_error.py	2009-11-08 19:17:54.000000000 +0700
+++ b/wimpiggy/test_error.py	2009-11-10 18:32:48.000000000 +0700
@@ -4,7 +4,7 @@
 # later version. See the file COPYING for details.
 
 from wimpiggy.test import *
-from wimpiggy.error import *
+from wimpiggy.error import trap, XError
 # Need a way to generate X errors...
 import wimpiggy.lowlevel
 
diff -ur a/wimpiggy/test_lowlevel.py b/wimpiggy/test_lowlevel.py
--- a/wimpiggy/test_lowlevel.py	2009-11-08 19:17:54.000000000 +0700
+++ b/wimpiggy/test_lowlevel.py	2009-11-10 18:33:15.000000000 +0700
@@ -6,7 +6,7 @@
 from wimpiggy.test import *
 import wimpiggy.lowlevel as l
 import gtk
-from wimpiggy.error import *
+from wimpiggy.error import trap, XError
 
 class TestLowlevel(TestWithSession):
     def root(self, disp=None):
diff -ur a/wimpiggy/test_selection.py b/wimpiggy/test_selection.py
--- a/wimpiggy/test_selection.py	2009-11-08 19:17:54.000000000 +0700
+++ b/wimpiggy/test_selection.py	2009-11-10 18:34:52.000000000 +0700
@@ -6,6 +6,7 @@
 from wimpiggy.test import *
 from wimpiggy.selection import ManagerSelection, AlreadyOwned
 import wimpiggy.lowlevel
+from wimpiggy.event import add_event_receiver
 
 import struct
 
@@ -46,7 +47,7 @@
         d2 = self.clone_display()
         root2 = d2.get_default_screen().get_root_window()
         root2.set_events(gtk.gdk.STRUCTURE_MASK)
-        wimpiggy.lowlevel.add_event_receiver(root2, self)
+        add_event_receiver(root2, self)
         d2.flush()
         self.event = None
 
diff -ur a/wimpiggy/world_window.py b/wimpiggy/world_window.py
--- a/wimpiggy/world_window.py	2009-11-08 19:17:54.000000000 +0700
+++ b/wimpiggy/world_window.py	2009-11-10 18:37:19.000000000 +0700
@@ -8,7 +8,7 @@
 import wimpiggy.lowlevel
 import wimpiggy.window
 import wimpiggy.prop
-from wimpiggy.error import *
+from wimpiggy.error import trap
 from wimpiggy.log import Logger
 log = Logger()
 
Only in b/working-notes/specs: .svn
Only in b/working-notes: .svn
diff -ur a/xpra/client.py b/xpra/client.py
--- a/xpra/client.py	2009-11-10 21:52:10.000000000 +0700
+++ b/xpra/client.py	2009-11-10 19:33:25.000000000 +0700
@@ -6,6 +6,8 @@
 import gtk
 import gobject
 import cairo
+import sys
+import hmac
 
 from wimpiggy.util import (gtk_main_quit_really,
                            gtk_main_quit_on_fatal_exceptions_enable)
diff -ur a/xpra/clipboard.py b/xpra/clipboard.py
--- a/xpra/clipboard.py	2009-11-08 19:17:54.000000000 +0700
+++ b/xpra/clipboard.py	2009-11-10 19:05:11.000000000 +0700
@@ -9,7 +9,7 @@
 
 from wimpiggy.util import n_arg_signal
 from wimpiggy.prop import prop_get
-from wimpiggy.error import *
+from wimpiggy.error import trap
 from wimpiggy.lowlevel import (get_xatom, get_pywindow,
                                gdk_atom_objects_from_gdk_atom_array)
 
diff -ur a/xpra/xsettings.py b/xpra/xsettings.py
--- a/xpra/xsettings.py	2009-11-08 19:17:54.000000000 +0700
+++ b/xpra/xsettings.py	2009-11-10 19:30:03.000000000 +0700
@@ -1,6 +1,6 @@
 import gobject
 import gtk
-from wimpiggy.error import *
+from wimpiggy.error import XError, trap
 from wimpiggy.util import no_arg_signal, one_arg_signal
 from wimpiggy.selection import ManagerSelection
 from wimpiggy.prop import prop_set, prop_get
@@ -84,7 +84,7 @@
         log("Fetching current XSettings data")
         try:
             return trap.call(self._get_settings_blob)
-        except XError, e:
+        except XError:
             log("X error while fetching XSettings data; ignored")
             return None
         
