diff -ur a/setup.py b/setup.py
--- a/setup.py	2009-11-08 19:17:54.000000000 +0700
+++ b/setup.py	2009-11-10 19:19:08.000000000 +0700
@@ -15,21 +15,8 @@
 
 from distutils.core import setup
 from distutils.extension import Extension
-from Pyrex.Distutils import build_ext
 import commands, os, sys
 
-from Pyrex.Compiler.Version import version as pyrex_version_string
-pyrex_version = [int(part) for part in pyrex_version_string.split(".")]
-# This was when the 'for 0 < i < 10:' syntax as added, bump upwards as
-# necessary:
-NEEDED_PYREX = (0, 9, 7)
-if tuple(pyrex_version) < NEEDED_PYREX:
-    sys.exit("ERROR: Your version of Pyrex is too old to build this package\n"
-             "You have version %s\n"
-             "Please upgrade to Pyrex %s or better"
-             % (pyrex_version_string,
-                ".".join([str(part) for part in NEEDED_PYREX])))
-
 # Tweaked from http://aspn.activestate.com/ASPN/Cookbook/Python/Recipe/502261
 def pkgconfig(*packages, **kw):
     flag_map = {'-I': 'include_dirs',
@@ -65,6 +52,36 @@
   xpra:
     %s""" % (wimpiggy_desc, parti_desc, xpra_desc)
 
+if not sys.platform.startswith("win"):
+    from Pyrex.Compiler.Version import version as pyrex_version_string
+    pyrex_version = [int(part) for part in pyrex_version_string.split(".")]
+    # This was when the 'for 0 < i < 10:' syntax as added, bump upwards as
+    # necessary:
+    NEEDED_PYREX = (0, 9, 7)
+    if tuple(pyrex_version) < NEEDED_PYREX:
+        sys.exit("ERROR: Your version of Pyrex is too old to build this package\n"
+             "You have version %s\n"
+             "Please upgrade to Pyrex %s or better"
+             % (pyrex_version_string,
+                ".".join([str(part) for part in NEEDED_PYREX])))
+    from Pyrex.Distutils import build_ext
+    ext_modules=[ 
+      Extension("wimpiggy.lowlevel.bindings",
+                ["wimpiggy/lowlevel/wimpiggy.lowlevel.bindings.pyx"],
+                **pkgconfig("pygobject-2.0", "gdk-x11-2.0", "gtk+-x11-2.0",
+                            "xtst", "xfixes", "xcomposite", "xdamage")
+                ),
+      Extension("xpra.wait_for_x_server",
+                ["xpra/xpra.wait_for_x_server.pyx"],
+                **pkgconfig("x11")
+                ),
+      ]
+    # Turn on Pyrex-sensitivity:
+    cmdclass = {'build_ext': build_ext}
+else:
+    ext_modules=None
+    cmdclass={}
+
 setup(
     name="parti-all",
     author="Nathaniel Smith",
@@ -82,17 +99,6 @@
              "scripts/xpra",
              ],
     data_files = [("share/man/man1", ["xpra.1"])],
-    ext_modules=[ 
-      Extension("wimpiggy.lowlevel.bindings",
-                ["wimpiggy/lowlevel/wimpiggy.lowlevel.bindings.pyx"],
-                **pkgconfig("pygobject-2.0", "gdk-x11-2.0", "gtk+-x11-2.0",
-                            "xtst", "xfixes", "xcomposite", "xdamage")
-                ),
-      Extension("xpra.wait_for_x_server",
-                ["xpra/xpra.wait_for_x_server.pyx"],
-                **pkgconfig("x11")
-                ),
-      ],
-    # Turn on Pyrex-sensitivity:
-    cmdclass = {'build_ext': build_ext}
+    ext_modules=ext_modules,
+    cmdclass = cmdclass
     )
