diff -ur a/xpra/client.py b/xpra/client.py
--- a/xpra/client.py	2009-11-10 21:57:32.000000000 +0700
+++ b/xpra/client.py	2009-11-10 21:59:09.000000000 +0700
@@ -86,10 +86,10 @@
     def update_metadata(self, metadata):
         self._metadata.update(metadata)
         
-        self.set_title(u"%s (via xpra)"
-                       % self._metadata.get("title",
+        self.set_title(u"%s%s"
+                       % (self._metadata.get("title",
                                             "<untitled window>"
-                                            ).decode("utf-8"))
+                                            ).decode("utf-8"), self._client.title_suffix))
 
         if "size-constraints" in self._metadata:
             size_metadata = self._metadata["size-constraints"]
diff -ur a/xpra/scripts/main.py b/xpra/scripts/main.py
--- a/xpra/scripts/main.py	2009-11-10 22:12:42.000000000 +0700
+++ b/xpra/scripts/main.py	2009-11-10 21:56:15.000000000 +0700
@@ -56,6 +56,9 @@
                       dest="bind_tcp", default=None,
                       metavar="[HOST]:PORT",
                       help="Listen for connections over TCP (insecure)")
+    parser.add_option("--title-suffix", action="store",
+                      dest="title_suffix", default=" (via xpra)",
+                      help="Text which is appended to the window's title")
     parser.add_option("-z", "--compress", action="store",
                       dest="compression_level", type="int", default=3,
                       metavar="LEVEL",

