--- a/xpra/scripts/main.py	2009-07-28 20:17:35.000000000 +0700
+++ b/xpra/scripts/main.py	2009-10-13 21:20:28.000000000 +0700
@@ -155,10 +167,16 @@
 
 def run_client(parser, opts, extra_args):
     from xpra.client import XpraClient
-    sock, local = client_sock(parser, opts, pick_display(parser, extra_args))
+    try:
+        sock, _ = client_sock(parser, opts, pick_display(parser, extra_args))
+    except socket.error, e:
+        sys.stdout.write("Failed to connect to socket: %s\n" % e)
+        return
     if opts.compression_level < 0 or opts.compression_level > 9:
         parser.error("Compression level must be between 0 and 9 inclusive.")
     if opts.jpegquality < 0 or opts.jpegquality > 100:
