### Eclipse Workspace Patch 1.0
#P xpra
Index: trunk/dev/wimpiggy/error.py
===================================================================
--- trunk/dev/wimpiggy/error.py	(revision 124)
+++ trunk/dev/wimpiggy/error.py	(working copy)
@@ -111,7 +111,7 @@
     def swallow_synced(self, fun, *args, **kwargs):
         try:
             self.call_synced(fun, *args, **kwargs)
-        except XError:
+        except XError, e:
             log("Ignoring X error: %s", e)
             pass
         return None
