### Eclipse Workspace Patch 1.0
#P xpra
Index: trunk/dev/parti/scripts/main.py
===================================================================
--- trunk/dev/parti/scripts/main.py	(revision 124)
+++ trunk/dev/parti/scripts/main.py	(working copy)
@@ -14,7 +14,7 @@
     parser.add_option("--replace", action="store_true",
                       dest="replace", default=False,
                       help="Replace any running window manager with Parti")
-    (options, args) = parser.parse_args(cmdline[1:])
+    (options, _) = parser.parse_args(cmdline[1:])
 
     # This means, if an exception propagates to the gtk mainloop, then pass it
     # on outwards.  Or at least it did at one time; dunno if it actually does
Index: trunk/dev/parti/scripts/repl.py
===================================================================
--- trunk/dev/parti/scripts/repl.py	(revision 124)
+++ trunk/dev/parti/scripts/repl.py	(working copy)
@@ -12,7 +12,7 @@
 
 def main(cmdline):
     parser = PartiOptionParser()
-    (options, args) = parser.parse_args(cmdline[1:])
+    parser.parse_args(cmdline[1:])
 
     # This means, if an exception propagates to the gtk mainloop, then pass it
     # on outwards.  Or at least it did at one time; dunno if it actually does
