--- upstream/xpra/server.py	2009-05-10 17:58:18.543358057 +0100
+++ dev/xpra/server.py	2009-05-10 17:58:00.411361055 +0100
@@ -247,6 +247,7 @@
         try:
             xmodmap = subprocess.Popen(["xmodmap", "-"], stdin=subprocess.PIPE)
         except OSError, e:
+            import sys
             sys.stderr.write("Error running xmodmap: %s\n" % (e,))
         xmodmap.communicate("""clear Lock
                                clear Shift
