--- parti-all-0.0.6-as-sent/setup.py	2009-03-22 16:34:43.000000000 +0700
+++ parti-all-0.0.6-dev/setup.py	2009-03-22 22:33:47.000000000 +0700
@@ -15,8 +15,7 @@
 
 from distutils.core import setup
 from distutils.extension import Extension
-from Pyrex.Distutils import build_ext
-import commands, os
+import commands, os, sys
 
 # Tweaked from http://aspn.activestate.com/ASPN/Cookbook/Python/Recipe/502261
 def pkgconfig(*packages, **kw):
@@ -53,6 +52,25 @@
   xpra:
     %s""" % (wimpiggy_desc, parti_desc, xpra_desc)
 
+if not sys.platform.startswith("win"):
+    from Pyrex.Distutils import build_ext
+    ext_modules=[ 
+      Extension("wimpiggy.lowlevel.bindings",
+                ["wimpiggy/lowlevel/wimpiggy.lowlevel.bindings.pyx"],
+                **pkgconfig("pygobject-2.0", "gdk-x11-2.0", "gtk+-x11-2.0",
+                            "xtst", "xfixes", "xcomposite", "xdamage")
+                ),
+      Extension("xpra.wait_for_x_server",
+                ["xpra/xpra.wait_for_x_server.pyx"],
+                **pkgconfig("x11")
+                ),
+      ]
+    # Turn on Pyrex-sensitivity:
+    cmdclass = {'build_ext': build_ext}
+else:
+    ext_modules=None
+    cmdclass={}
+
 setup(
     name="parti-all",
     author="Nathaniel Smith",
@@ -70,17 +88,6 @@
              "scripts/xpra",
              ],
     data_files = [("share/man/man1", ["xpra.1"])],
-    ext_modules=[ 
-      Extension("wimpiggy.lowlevel.bindings",
-                ["wimpiggy/lowlevel/wimpiggy.lowlevel.bindings.pyx"],
-                **pkgconfig("pygobject-2.0", "gdk-x11-2.0", "gtk+-x11-2.0",
-                            "xtst", "xfixes", "xcomposite", "xdamage")
-                ),
-      Extension("xpra.wait_for_x_server",
-                ["xpra/xpra.wait_for_x_server.pyx"],
-                **pkgconfig("x11")
-                ),
-      ],
-    # Turn on Pyrex-sensitivity:
-    cmdclass = {'build_ext': build_ext}
+    ext_modules=ext_modules,
+    cmdclass = cmdclass
     )
