--- /home/antoine/projects/eclipse-workspaces/workspace-shifter/xpra/trunk/dev/xpra/scripts/main.py	2009-06-15 01:12:16.000000000 +0700
+++ ./xpra/scripts/main.py	2009-06-15 01:15:17.000000000 +0700
@@ -92,9 +92,7 @@
 
     mode = args.pop(0)
     if mode in ("start", "upgrade"):
-        nox()
-        from xpra.scripts.server import run_server
-        run_server(parser, options, mode, script_file, args)
+        print "server disabled on windows"
     elif mode == "attach":
         run_client(parser, options, args)
     elif mode == "stop":
