--- parti-all-0.0.7-pre-vanilla/xpra/client.py	2009-04-03 08:29:09.000000000 +0200
+++ parti-all-0.0.7-pre/xpra/client.py	2009-05-04 21:46:45.430683756 +0200
@@ -8,6 +8,7 @@
 import cairo
 import os
 import os.path
+import sys
 
 from wimpiggy.util import (one_arg_signal,
                            gtk_main_quit_really,
@@ -19,7 +20,6 @@
 
 from xpra.protocol import Protocol
 from xpra.keys import mask_to_names
-from xpra.clipboard import ClipboardProtocolHelper
 
 import xpra
 default_capabilities = {"__prerelease_version": xpra.__version__}
@@ -284,7 +285,11 @@
         self._keymap.connect("keys-changed", self._keys_changed)
         self._keys_changed()
 
-        self._clipboard_helper = ClipboardProtocolHelper(self.send)
+        if not sys.platform.startswith("win"):
+            from xpra.clipboard import ClipboardProtocolHelper
+            self._clipboard_helper = ClipboardProtocolHelper(self.send)
+        else:
+            self._clipboard_helper = None    
 
         self._focused = None
 
@@ -322,7 +327,8 @@
         if capabilities.get("__prerelease_version") != xpra.__version__:
             log.error("sorry, I only know how to talk to v%s servers",
                       xpra.__version__)
-        self._clipboard_helper.send_all_tokens()
+        if self._clipboard_helper:
+            self._clipboard_helper.send_all_tokens()
 
     def _process_new_common(self, packet, override_redirect):
         (_, id, x, y, w, h, metadata) = packet
@@ -381,7 +387,8 @@
         packet_type = packet[0]
         if (isinstance(packet_type, str)
             and packet_type.startswith("clipboard-")):
-            self._clipboard_helper.process_clipboard_packet(packet)
+            if self._clipboard_helper:
+                self._clipboard_helper.process_clipboard_packet(packet)
         else:
             self._packet_handlers[packet_type](self, packet)
 
