Index: trunk/dev/xpra/dotxpra.py
===================================================================
--- a/xpra/dotxpra.py	(revision 124)
+++ b/xpra/dotxpra.py	(working copy)
@@ -3,7 +3,6 @@
 # Parti is released under the terms of the GNU GPL v2, or, at your option, any
 # later version. See the file COPYING for details.
 
-import os
 import os.path
 import socket
 import errno
@@ -42,6 +41,7 @@
         if not os.path.exists(path):
             return self.DEAD
         sock = socket.socket(socket.AF_UNIX)
+        sock.settimeout(5)
         try:
             sock.connect(path)
         except socket.error, e:
