Index: dev/xpra/darwin/gui.py
===================================================================
--- dev/xpra/darwin/gui.py	(revision 301)
+++ dev/xpra/darwin/gui.py	(working copy)
@@ -41,7 +41,7 @@
             self.send(["clipboard-contents-none", request_id, selection])
 
 class ClientExtras(object):
-    def __init__(self, send_packet_cb):
+    def __init__(self, send_packet_cb, pulseaudio):
         self.send = send_packet_cb
         self.setup_macdock()
 
Index: dev/xpra/xposix/gui.py
===================================================================
--- dev/xpra/xposix/gui.py	(revision 293)
+++ dev/xpra/xposix/gui.py	(working copy)
@@ -14,8 +14,15 @@
 from xpra.xposix.xroot_props import XRootPropWatcher
 
 class ClientExtras(object):
-    def __init__(self, send_packet_cb):
+    def __init__(self, send_packet_cb, pulseaudio):
         self.send = send_packet_cb
+        self.ROOT_PROPS = {
+            "RESOURCE_MANAGER": "resource-manager"
+            }
+        if pulseaudio:
+            self.ROOT_PROPS["PULSE_COOKIE"] = "pulse-cookie"
+            self.ROOT_PROPS["PULSE_ID"] = "pulse-id"
+            self.ROOT_PROPS["PULSE_SERVER"] = "pulse-server"
 
     def handshake_complete(self, capabilities):
         self._xsettings_watcher = XSettingsWatcher()
@@ -32,13 +39,6 @@
         if blob is not None:
             self.send(["server-settings", {"xsettings-blob": blob}])
 
-    ROOT_PROPS = {
-        "RESOURCE_MANAGER": "resource-manager",
-        "PULSE_COOKIE": "pulse-cookie",
-        "PULSE_ID": "pulse-id",
-        "PULSE_SERVER": "pulse-server",
-        }
-    
     def _handle_root_prop_changed(self, obj, prop, value):
         assert prop in self.ROOT_PROPS
         if value is not None:
Index: dev/xpra/win32/gui.py
===================================================================
--- dev/xpra/win32/gui.py	(revision 293)
+++ dev/xpra/win32/gui.py	(working copy)
@@ -41,7 +41,7 @@
             self.send(["clipboard-contents-none", request_id, selection])
 
 class ClientExtras(object):
-    def __init__(self, send_packet_cb):
+    def __init__(self, send_packet_cb, pulseaudio):
         self.send = send_packet_cb
 
     def handshake_complete(self, capabilities):
Index: dev/xpra/client.py
===================================================================
--- dev/xpra/client.py	(revision 293)
+++ dev/xpra/client.py	(working copy)
@@ -307,7 +307,7 @@
 
         # FIXME: these should perhaps be merged.
         self._clipboard_helper = ClipboardProtocolHelper(self.send)
-        self._client_extras = ClientExtras(self.send)
+        self._client_extras = ClientExtras(self.send, pulseaudio)
 
         self._focused = None
 
