diff -ur a/xpra/client.py b/xpra/client.py
--- a/xpra/client.py	2009-11-08 19:17:54.000000000 +0700
+++ b/xpra/client.py	2009-11-10 19:33:25.000000000 +0700
@@ -16,9 +18,6 @@
 
 from xpra.protocol import Protocol
 from xpra.keys import mask_to_names
-from xpra.clipboard import ClipboardProtocolHelper
-from xpra.xsettings import XSettingsWatcher
-from xpra.root_props import RootPropWatcher
 from wimpiggy.prop import prop_get
 
 import xpra
@@ -289,7 +294,11 @@
         self._xsettings_watcher = None
         self._root_props_watcher = None
 
-        self._clipboard_helper = ClipboardProtocolHelper(self.send)
+        if not sys.platform.startswith("win"):
+            from xpra.clipboard import ClipboardProtocolHelper
+            self._clipboard_helper = ClipboardProtocolHelper(self.send)
+        else:
+            self._clipboard_helper = None    
 
         self._focused = None
 
@@ -359,16 +389,22 @@
                          "Please complain to "
                          "parti-discuss@partiwm.org"
                          % (avail_w, avail_h, root_w, root_h))
-        self._clipboard_helper.send_all_tokens()
-        self._xsettings_watcher = XSettingsWatcher()
-        self._xsettings_watcher.connect("xsettings-changed",
-                                        self._handle_xsettings_changed)
-        self._handle_xsettings_changed()
-        self._root_props_watcher = RootPropWatcher(self.ROOT_PROPS.keys())
-        self._root_props_watcher.connect("root-prop-changed",
-                                        self._handle_root_prop_changed)
-        for prop in self.ROOT_PROPS:
-            self._handle_root_prop_changed(None, prop)
+        if self._clipboard_helper:
+            self._clipboard_helper.send_all_tokens()
+        self._xsettings_watcher = None
+        self._root_props_watcher = None
+        if not sys.platform.startswith("win"):
+            from xpra.xsettings import XSettingsWatcher
+            from xpra.root_props import RootPropWatcher
+            self._xsettings_watcher = XSettingsWatcher()
+            self._xsettings_watcher.connect("xsettings-changed",
+	                                        self._handle_xsettings_changed)
+            self._handle_xsettings_changed()
+            self._root_props_watcher = RootPropWatcher(self.ROOT_PROPS.keys())
+            self._root_props_watcher.connect("root-prop-changed",
+	                                        self._handle_root_prop_changed)
+            for prop in self.ROOT_PROPS:
+                self._handle_root_prop_changed(None, prop)
 
     def _process_new_common(self, packet, override_redirect):
         (_, id, x, y, w, h, metadata) = packet
@@ -427,7 +463,8 @@
         packet_type = packet[0]
         if (isinstance(packet_type, str)
             and packet_type.startswith("clipboard-")):
-            self._clipboard_helper.process_clipboard_packet(packet)
+            if self._clipboard_helper:
+                self._clipboard_helper.process_clipboard_packet(packet)
         else:
             self._packet_handlers[packet_type](self, packet)
 
--- as-sent/xpra/client.py	2009-11-10 23:03:48.000000000 +0700
+++ ./dev/xpra/client.py	2009-11-14 16:18:56.000000000 +0700
@@ -18,7 +18,6 @@
 
 from xpra.protocol import Protocol
 from xpra.keys import mask_to_names
-from wimpiggy.prop import prop_get
 
 import xpra
 default_capabilities = {"__prerelease_version": xpra.__version__}
@@ -365,6 +364,7 @@
         }
     
     def _handle_root_prop_changed(self, obj, prop):
+        from wimpiggy.prop import prop_get
         assert prop in self.ROOT_PROPS
         v = prop_get(gtk.gdk.get_default_root_window(),
                      prop, "latin1")
