--- as-sent/xpra/scripts/main.py	2009-11-10 23:03:48.000000000 +0700
+++ dev/xpra/scripts/main.py	2009-11-11 00:57:53.000000000 +0700
@@ -188,10 +188,12 @@
     elif display_desc["type"] == "unix-domain":
         sockdir = DotXpra()
         sock = socket.socket(socket.AF_UNIX)
+        sock.settimeout(10)
         sock.connect(sockdir.socket_path(display_desc["display"]))
         return sock
     elif display_desc["type"] == "tcp":
         sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        sock.settimeout(20)
         sock.connect((display_desc["host"], display_desc["port"]))
         return sock
     else:
